pkg/dbutil/dbutil.go

```diff
@@ -0,0 +1,299 @@
+
+	if err := tx.PersistLayer(layer.Hash, layer.Features, layer.Namespaces, layer.By); err != nil {
+		return err
+	}
+
+	return tx.Commit()
+}
+
+// PersistFeatures wraps session PersistFeatures function with begin and commit.
+func PersistFeatures(datastore database.Datastore, features []database.Feature) error {
+	tx, err := datastore.Begin()
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
+
+	if err := tx.PersistFeatures(features); err != nil {
+		return err
+	}
+	return tx.Commit()
+}
+
+// PersistNamespaces wraps session PersistNamespaces function with begin and
+// commit.
+func PersistNamespaces(datastore database.Datastore, namespaces []database.Namespace) error {
+	tx, err := datastore.Begin()
+	if err != nil {
+		return err
+	}
+	defer tx.Rollback()
+
+	if err := tx.PersistNamespaces(namespaces); err != nil {
+		return err
+	}
+
+	return tx.Commit()
+}
+
+// FindAncestry wraps session FindAncestry function with begin and rollback.
+func FindAncestry(datastore database.Datastore, name string) (database.Ancestry, bool, error) {
+	tx, err := datastore.Begin()
+	defer tx.Rollback()
+
+	if err != nil {
+		return database.Ancestry{}, false, err
+	}
+
+	return tx.FindAncestry(name)
+}
+
+func FindLayer(datastore database.Datastore, hash string) (layer database.Layer, ok bool, err error) {
+	var tx database.Session
+	if tx, err = datastore.Begin(); err != nil {
+		return
+	}
+
+	defer tx.Rollback()
+	layer, ok, err = tx.FindLayer(hash)
+	return
+}
+
+func DeduplicateNamespacedFeatures(features []database.NamespacedFeature) []database.NamespacedFeature {
+	nsSet := mapset.NewSet()
+	for _, ns := range features {
+		nsSet.Add(ns)
+	}
+
+	result := make([]database.NamespacedFeature, 0, nsSet.Cardinality())
+	for ns := range nsSet.Iter() {
+		result = append(result, ns.(database.NamespacedFeature))
+	}
+
+	return result
+}
+
+func GetAncestryFeatures(ancestry database.Ancestry) []database.NamespacedFeature {
+	features := []database.NamespacedFeature{}
+	for _, layer := range ancestry.Layers {
+		features = append(features, layer.GetFeatures()...)
+	}
+
+	return DeduplicateNamespacedFeatures(features)
+}
+
+func UpsertAncestry(datastore database.Datastore, ancestry database.Ancestry) error {
+	tx, err := datastore.Begin()
+	if err != nil {
+		return err
+	}
+
+	if err = tx.UpsertAncestry(ancestry); err != nil {
+		tx.Rollback()
+		return err
+	}
+
+	if err = tx.Commit(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// PersistNamespacedFeatures saves features to database.
+func PersistNamespacedFeatures(datastore database.Datastore, features []database.NamespacedFeature) error {
+	tx, err := datastore.Begin()
+	if err != nil {
+		return err
+	}
+
+	if err := tx.PersistNamespacedFeatures(features); err != nil {
+		tx.Rollback()
+		return err
+	}
+
+	if err := tx.Commit(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func CacheRelatedVulnerability(datastore database.Datastore, features []database.NamespacedFeature) error {
```

Comment:
TODO: I start to think that the abstraction on database layer is wrong. Somehow we need to figure out a way to get ride of the implicit database feature relationship.